File: 112/path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, sum int) bool {
	if nil == root {
		return false
	}

	if nil == root.Left && nil == root.Right {
		if root.Val == sum {
			return true
		}
		return false
	}

	var hasLeftPath, hasRightPath bool
	if root.Left != nil {
		hasLeftPath = hasPathSum(root.Left, sum-root.Val)
	}
	if root.Right != nil {
		hasRightPath = hasPathSum(root.Right, sum-root.Val)
	}
	return hasLeftPath || hasRightPath
}

File: 112/path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {boolean}
 */
var hasPathSum = function (root, sum) {
    if (!root)
        return false;

    if (!root.left && !root.right) {
        if (root.val == sum)
            return true;
        return false;
    }

    let hasLeftPath, hasRightPath;
    if (root.left)
        hasLeftPath = hasPathSum(root.left, sum - root.val);

    if (root.right)
        hasRightPath = hasPathSum(root.right, sum - root.val);

    return !!(hasLeftPath || hasRightPath);
};